File: 1760-Minimum Limit of Balls in a Bag.py
class Solution:
    def minimumSize(self, nums: List[int], maxOperations: int) -> int:
        
        left, right = 1, max(nums)
        k = maxOperations
        
        def achieve(target):
            limit = k
            for num in nums:
                if num > target:
                    limit -= (num+target-1)//target-1
                    if limit < 0:
                        return False
            return True
        
        while left + 1 < right:
            mid = left + (right - left)//2
            if achieve(mid):
                right = mid
            else:
                left = mid
                
        if achieve(right):
            return right
        else:
            return left

File: C++/1760-Minimum Limit of Balls in a Bag.cpp
class Solution {
public:
    bool isOK(vector<int>& nums, int maxOperations, int smallest) {
        int op = 0;
        for (int num:nums) {
            int diff = num - smallest;
            if (diff <= 0) continue;
            op += (diff/smallest) + (diff%smallest?1:0);
            if (op > maxOperations) return false;
        }
        return true;
    }
    int minimumSize(vector<int>& nums, int maxOperations) {
        int left = 1, right = *max_element(begin(nums), end(nums));
        while (left + 1 < right) {
            int mid = left + (right-left)/2;
            if (isOK(nums, maxOperations, mid)) {
                right = mid;
            } else {
                left = mid;
            }
        }
        if (isOK(nums, maxOperations, right)) {
            return right;
        }
        return left;
    }
};